repair/row_level.cc

```diff
@@ -53,6 +53,82 @@
 
 extern logging::logger rlogger;
 
+class token_bucket {
+private:
+    size_t _fill_rate;
+    size_t _max_tokens;
+    size_t _current_tokens;
+    seastar::condition_variable _cond;
+    seastar::timer<std::chrono::steady_clock> _fill_timer;
+    std::chrono::milliseconds _fill_interval{10};
+private:
+    void fill() {
+        auto tokens = _fill_rate * _fill_interval.count() / std::chrono::milliseconds(1000).count();
+        _current_tokens = std::min(_max_tokens, _current_tokens + tokens);
+        _cond.broadcast();
+    }
+    bool enabled() {
+        return _fill_rate != 0;
+    }
+public:
+    token_bucket(size_t fill_rate, size_t max_tokens)
+        : _fill_rate(fill_rate)
+        , _max_tokens(max_tokens)
+        , _current_tokens(max_tokens) {
+        _fill_timer.set_callback([this] { fill(); });
+        set_rate(fill_rate);
+    }
+    future<> consume(size_t tokens) {
+        if (!enabled()) {
+            return make_ready_future<>();
+        }
+        if (tokens > _max_tokens) {
+            rlogger.warn("token_bucket: tokens={} requested are bigger than max_tokens={}", tokens, _max_tokens);
+            return make_ready_future<>();
+        }
+        return _cond.wait([this, tokens] { return tokens <= _current_tokens || !enabled(); }).then([this, tokens] {
+            if (enabled()) {
+                _current_tokens -= tokens;
+            } else {
+                _current_tokens = _max_tokens;
+            }
```

Comment:
Indeed, there is a race because the check and the decreasing the tokens are not atomic. I think we can solve this by moving them together. 

````
  81     future<> consume(size_t tokens) {
  82         if (!enabled()) {
  83             return make_ready_future<>();
  84         }
  85         if (tokens > _max_tokens) {
  86             rlogger.warn("token_bucket: tokens={} requested are bigger than max_tokens={}", tokens, _max_tokens);
  87             return make_ready_future<>();
  88         }
  89         return _cond.wait([this, tokens] {
  90             bool wake_up = tokens <= _current_tokens || !enabled();
  91             if (wake_up) {
  92                 if (enabled()) {
  93                     _current_tokens -= tokens;
  94                 } else {
  95                     _current_tokens = _max_tokens;
  96                 }
  97             }
  98             return wake_up;
  99         });
 100     }
```